### PR TITLE
Prevent reshard concurrent with calling stop

### PR DESCRIPTION
**Pull request by vsliouniaev**
We should wait util the reshard loop and reshard calculations are complete before calling stop, otherwise the reshard stop, start can happen concurrently.

In practice this happens when doing a Prometheus reload at the same time as a reshard loop is happening

fixes: https://github.com/prometheus/prometheus/issues/5451

### CODE REVIEW

**Comment by brian-brazil** [comment]
@cstyan You were looking into this.

File: storage/remote/queue_manager.go [hunk 1]
```diff
@@ -322,9 +322,9 @@ func (t *QueueManager) Stop() {
 	defer level.Info(t.logger).Log("msg", "Remote storage stopped.")
 
 	close(t.quit)
+	t.wg.Wait()
```
Comment (cstyan) [inline storage/remote/queue_manager.go]: Can we leave a comment here? Something like
```suggestion
	// Wait for all QueueManager routines to end before stopping shards and WAL watcher. This
	// is to ensure we don't end up executing a reshard and shards.stop() at the same time, which
	// causes a closed channel panic.
```

**Comment by vsliouniaev** [comment]
I initially also tried a locking approach, but depending on where locks were added, I would end up with hard to reproduce deadlocks, due to the comms all being done through asynchronous channels.

**Comment by brian-brazil** [comment]
This looks like the right fix to me. Thanks!

**Pull request merged**